src/librustc/mir/mod.rs

```diff
@@ -3050,6 +3058,11 @@ pub struct GeneratorLayout<'tcx> {
     /// be stored in multiple variants.
     pub variant_fields: IndexVec<VariantIdx, IndexVec<Field, GeneratorSavedLocal>>,
 
+    /// Which saved locals are storage-live at the same time. Locals that do not
+    /// have conflicts with each other are allowed to overlap in the computed
+    /// layout.
+    pub storage_conflicts: IndexVec<GeneratorSavedLocal, BitSet<GeneratorSavedLocal>>,
```

Comment:
FWIW, storing this in MIR, instead of the "which `GeneratorSavedLocal` is eligible for going into per-variant data" information, means layout has to redo work for every monomorphization of a generator.

Reply:
You could have pretty much just `conflicting_locals: BitSet<GeneratorSavedLocal>` and "present in only one variant and not in `conflicting_locals`" would give "is this `GeneratorSavedLocal` eligible?".

Reply:
Mm yeah, we are redoing unnecessary work right now. I was imagining that we'd replace the layout computation with a smarter one in the future which does depend on the inner types. I don't really like the idea of factoring the code to make that harder..

Reply:
This should probably be named `liveness_conflicts` and not refer to storage-liveness. We could use any conservative estimation of liveness here.

Reply:
Storage liveness is pretty much what we want I'd say. I can remove the explicit mention of storage-live from the comment, but I think the name is pretty good, given that we are deciding when to overlap the storage of these fields.

_I think_ we have all the same constraints as storage liveness, in the sense that if we make the liveness estimation used for this optimization less conservative, it must also be valid for storage liveness, and vice versa. For example, it's not valid to move up the `StorageDead` of a local that has a live borrow, and that would also break the optimization.

I'm imagining that if we ever do "tighten" the liveness estimation, we do it as a MIR pass using `StorageLive`/`StorageDead` for this reason. That would have the additional benefit of optimizing codegen downstream, I think. Though in practice, maybe we won't do this for performance reasons.